lib/trino-memory-context/src/main/java/io/trino/memory/context/ValidatingLocalMemoryContext.java

```diff
@@ -0,0 +1,108 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.memory.context;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.log.Logger;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class ValidatingLocalMemoryContext
+        implements LocalMemoryContext
+{
+    private static final Logger log = Logger.get(ValidatingLocalMemoryContext.class);
+
+    private final LocalMemoryContext delegate;
+    private final String allocationTag;
+    private final MemoryAllocationValidator memoryValidator;
+
+    public ValidatingLocalMemoryContext(LocalMemoryContext delegate, String allocationTag, MemoryAllocationValidator memoryValidator)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.allocationTag = requireNonNull(allocationTag, "allocationTag is null");
+        this.memoryValidator = requireNonNull(memoryValidator, "memoryValidator is null");
+    }
+
+    @Override
+    public long getBytes()
+    {
+        return delegate.getBytes();
+    }
+
+    @Override
+    public ListenableFuture<Void> setBytes(long bytes)
+    {
+        long delta = bytes - delegate.getBytes();
+
+        // first consult validator if allocation is possible
+        memoryValidator.reserveMemory(allocationTag, delta);
+
+        // update the parent before updating usedBytes as it may throw a runtime exception (e.g., ExceededMemoryLimitException)
+        try {
+            // do actual allocation
+            ListenableFuture<Void> future = delegate.setBytes(bytes);
+            return future;
+        }
+        catch (Exception e) {
+            revertReservationInValidator(allocationTag, delta, Optional.of(e));
+            throw e;
+        }
+    }
+
+    @Override
+    public boolean trySetBytes(long bytes)
+    {
+        long delta = bytes - delegate.getBytes();
+
+        if (!memoryValidator.tryReserveMemory(allocationTag, delta)) {
+            return false;
+        }
+
+        try {
+            if (delegate.trySetBytes(bytes)) {
+                return true;
+            }
+        }
+        catch (Exception e) {
+            revertReservationInValidator(allocationTag, delta, Optional.of(e));
+            throw e;
+        }
+
+        revertReservationInValidator(allocationTag, delta, Optional.empty());
+        return false;
+    }
+
+    @Override
+    public void close()
+    {
+        delegate.close();
+    }
+
+    private void revertReservationInValidator(String allocationTag, long delta, Optional<Exception> revertCause)
+    {
+        try {
+            memoryValidator.reserveMemory(allocationTag, -delta);
+        }
+        catch (Exception suppressed) {
+            log.warn(suppressed, "Could not rollback memory reservation within allocation validator");
+            revertCause.ifPresent(cause -> {
```

Comment:
you consume `suppressed` when `revertCause` is empty.

Reply:
Changed to have two methods:
```
private void revertReservationInValidatorSuppressing(String allocationTag, long delta, Exception revertCause)
``` 

and 

```
private void revertReservationInValidator(String allocationTag, long delta)
```